### PR TITLE
Save Xref warnings to manifest and return diagnostics from compiler

### PR DESCRIPTION
**Pull request by JakeBecker**

### CODE REVIEW

File: lib/mix/lib/mix/tasks/compile.xref.ex [hunk 1]
```diff
@@ -28,28 +29,34 @@ defmodule Mix.Tasks.Compile.Xref do
...
     !!opts[:force] or Mix.Utils.stale?(E.manifests(), manifests())
   end
 
-  defp should_exit?(:error, opts),
-    do: warnings_as_errors(opts) == true
-  defp should_exit?(_, _opts),
-    do: false
+  defp run_xref do
+    timestamp = :calendar.universal_time()
+    warnings = Mix.Task.run("xref", ["warnings"])
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/compile.xref.ex]: Let's make it return `{:ok, warnings}` (and then `:noop` if it already ran) so it forces us to properly pattern match.

File: lib/mix/lib/mix/tasks/xref.ex [hunk 2]
```diff
@@ -129,11 +129,7 @@ defmodule Mix.Tasks.Xref do
   ## Modes
 
   defp warnings() do
-    if unreachable(&print_warnings/2) == [] do
-      :ok
-    else
-      :error
-    end
+    List.flatten(unreachable(&warnings/2))
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/xref.ex]: Do we really need `List.flatten`? `List.flatten` is almost always a smell. It means we were supposed to flat map or concat at some point before but we did not.

File: lib/mix/lib/mix/tasks/compile.xref.ex [hunk 3]
```diff
@@ -73,4 +92,18 @@ defmodule Mix.Tasks.Compile.Xref do
   def clean do
     File.rm manifest()
   end
+
+  defp to_diagnostics(warnings, severity) do
+    Enum.flat_map(warnings, fn {file, lines, message} ->
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/compile.xref.ex]: Using comprehensions all the way through will be clearer:

```elixir
for {file, lines, message} <- warnings,
    line <- lines do
  %Mix...
```

**Comment by JakeBecker** [comment]
@josevalim I made the changes. Thanks!

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**